server/src/main/java/org/opensearch/action/search/TransportSearchAction.java

```diff
@@ -121,6 +123,19 @@
     public static final Setting<Long> SHARD_COUNT_LIMIT_SETTING = Setting.longSetting(
             "action.search.shard_count.limit", Long.MAX_VALUE, 1L, Property.Dynamic, Property.NodeScope);
 
+    // cluster level setting for timeout based search cancellation. If search request level parameter is present then that will take
+    // precedence over the cluster setting value
+    public static final String SEARCH_REQUEST_CANCEL_AFTER_TIMEINTERVAL_SETTING_KEY = "search.cancel_after_timeinterval";
+    public static final Setting<TimeValue> SEARCH_REQUEST_CANCEL_AFTER_TIMEINTERVAL_SETTING =
+        Setting.timeSetting(SEARCH_REQUEST_CANCEL_AFTER_TIMEINTERVAL_SETTING_KEY, TimeValue.timeValueSeconds(300), SearchService.NO_TIMEOUT,
+            Setting.Property.Dynamic, Setting.Property.NodeScope);
+
+    // cluster level setting to control enabling/disabling the timeout based cancellation. This is enabled by default
+    public static final String SEARCH_REQUEST_CANCELLATION_ENABLE_SETTING_KEY = "search.timeout.cancellation.enable";
+    public static final Setting<Boolean> SEARCH_REQUEST_CANCELLATION_ENABLE_SETTING =
+        Setting.boolSetting(SEARCH_REQUEST_CANCELLATION_ENABLE_SETTING_KEY, true, Setting.Property.Dynamic,
+            Setting.Property.NodeScope);
```

Comment:
Yes we can do that, but my motivation with boolean is to support another case. Let's say the client applications are using the request level parameter and now there is some regression which is seen. Then it may not be possible or hard to change the application to avoid setting that parameter at request level. With the parameter being present in a request, cancellation will kick in for those. However, with boolean we get a switch to disable this feature entirely irrespective of valid parameter being set at request level or at cluster level. Thoughts ?

Reply:
> However, with boolean we get a switch to disable this feature entirely irrespective of valid parameter being set at request level or at cluster level

You could just as easily set the param to `null` or (even easier actually) remove it altogether if there is an issue. 

> Also the boolean for cluster setting and boolean in CancellableTask class have different purposes as described in the comments.

The cluster level setting is like a feature flag here. I think that should be managed on the client side. If there is an issue, handle your request by setting them all to a timeInterval of -1. 
On a maintainability level, this is an extra cluster setting that is just not a must-have.

Reply:
> > However, with boolean we get a switch to disable this feature entirely irrespective of valid parameter being set at request level or at cluster level
> 
> You could just as easily set the param to `null` or (even easier actually) remove it altogether if there is an issue.
> 
> > Also the boolean for cluster setting and boolean in CancellableTask class have different purposes as described in the comments.
> 
> The cluster level setting is like a feature flag here. I think that should be managed on the client side. If there is an issue, handle your request by setting them all to a timeInterval of -1.
> On a maintainability level, this is an extra cluster setting that is just not a must-have.

Yes, we can achieve that by setting request level parameter to null or -1. My understanding is to do that on client side usually it may need deployment or code change in client application. Also there may be multiple client applications with different request level timeout hitting the same cluster, so all such applications needs to be modified. Whereas, with cluster setting it is a single point of control and is updatable dynamically. Given the change touches the `SearchRequest` I will still incline on having this flag in V1 and I can create a tracking issue to deprecate after 2 or 3 releases ?

Reply:
Since we are developing upstream - the FF approach is not required since there is no notion of temporary settings. In fact, removing settings would kind of be a breaking change which we would love to avoid as much as possible.
Regarding the ability to have such a FF - this is up to the user if they want to disable all. They could also as you pointed out in another comment - set the cluster setting interval value to -1.

Reply:
Probably I was not clear on how disable will work. Let me create a table to describe different scenarios and behavior. Based on below, the cluster level setting to -1 alone will not disable the feature. Request level param needs to be change for it.

enable flag |  request param | cluster param | Effective |  Feature   |
----------- |---------------- |---------------|-----------|---------|
   True          |        null                |       10s            |     10s      | |
   True          |        null                |       -1              |     -1         |  disabled |
   True          |          -1                |       10s           |     -1          |  disabled |
   True          |         20s              |       10s           |     20s       | |
   False       |          any               |       any           |     N/A       |  disabled |

On thinking more, currently we favor the request parameter over cluster one. We can repurpose the cluster level boolean setting to enforce precedence instead. Something like `enforce.server.search.cancellation: true/false`. This will have 2 benefits over current boolean flag. 

1. Provide mechanism for disabling it completely from server side.
2. Provide mechanism to enforce bounds for non-behaving client which may abuse the request level parameter. This will help for containing the resource consumption by such clients.

With new setting below will be the behavior:

enforce flag |  request param | cluster param | Effective |  Feature   |
----------- |---------------- |---------------|-----------|---------|
   False          |        null                |       10s            |     10s      | |
   False          |        null                |       -1              |     -1         |  disabled |
   False          |          -1                |       10s           |     -1          |  disabled |
   False          |         20s              |       10s           |     20s       | |
   True       |          null               |       10s           |     10s       |   |
   True       |          null               |       -1           |     -1       |  disabled |
   True       |          -1               |       10s           |     10s       |   |
   True       |          20s               |       10s           |     10s       |   |

Reply:
@AmiStrn - Changes were trivial, so I have made the change for new cluster setting in latest commit itself. Will update the description of PR to reflect the same.

Reply:
> Since we are developing upstream - the FF approach is not required since there is no notion of temporary settings. In fact, removing settings would kind of be a breaking change which we would love to avoid as much as possible.

Why is Feature Flags considered bad. eg. allocation service/circuit breakers have flags like `cluster.routing.allocation.enable`/`indices.breaker.total.use_real_memory`. It gives fine grained knobs to control and experiment with new features?

Reply:
@malpani I don't think Feature Flags considered are bad. I use them a lot! But when I do it is in CI/CD mostly. 
I think the major issue is the wording - Feature Flag is something I regard as temporary until a feature is released, then it is removed (without breaking anything). 
My comment is refering to this:
> I will still incline on having this flag in V1 and I can create a tracking issue to deprecate after 2 or 3 releases 

I think planning to create a breaking change is only ok for when we have no choice. I don't think this is propper use of Feature Flags that are meant to be dynamic and very short-lived. There are many settings that work this way as you point out, I don't like those either:) 
Most of those cases could be turned into plugins. This way users could install them and have a dedicated API to control them rather than via the cluster settings api.